### DISCUSSION TITLE
USR TCP232 304 - no data to poll

### DISCUSSION THREAD

**Discussion by ccbeorn** [nathanmarlor/foxess_modbus#417] [Support] [answered]
Hi everyone,

first and foremost: Thanks so much for the contribution of this integration HowTo, so much appreciated!!

I followed the Howto up until the wiring part and integration of the Adapter. I went for the TCP Adapter as mentioned in the subject line.

I am failing to get any data from the device and am now trying to figure out where the chain is broken. Whether the inverter does not (or can not) send the RS485 data to the adapter or whether I misconfigured something between the Adapter and my home network. 

I used two modbus tools to check for modbus data on the network but continue to get timeouts.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/81196830/944768ef-c32c-4b63-9532-f268ff4572db)

![image](https://github.com/nathanmarlor/foxess_modbus/assets/81196830/8a55e38f-aa00-417f-91f8-c611ab0dc51c)

Now, there is one fundamental thing puzzling me: that is the serial port configuration on the adapter. Note that my home network is 172.16.66.x 
I would have thought that this should be absolutely ok.
But why is there a fix "192.168.0.201" IP address configured for the work mode "TCP Server" (which I am supposed to use according to the howto)? For the RS485 side, this does not make any sense and most probably does not apply. But on the other side (the TCP side), it would screw up my connection of course if the adapter uses a 192... network where my home network is on 172...
Or can I completely ignore that IP?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/81196830/a0d50ff6-6a1d-491e-b6ef-9c84403006c9)

Note that I did another test and used a laptop with a 192.168.0.xxx address on the network switch where the adapter is connected. In that case the modpoll Tool states that it cannot reach the server, which seems to indicate that there is indeed no 192.168.0.201 being actively used on that side, so I can probably ignore the TCP Server IP that is listed in the serial config. (that might be the reason why this config element is blacked out in the screenshot of the howto),

Back to my main setup (using the 172.16.66.x network):
When I connect with the GUI based modpoll Tool, I can see on the adapter, that I am connected apparently, still, I receive only timeouts, no matter what I try to read.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/81196830/804af983-599c-4bf6-b006-0f1f4903e435)

The 172.16.66.6 belongs to my main computer on which I use the mod poll tool.

Do I interpret this correctly, that the connection is established but there is just nothing to read because the inverter is probably not providing anything on the RS485 side?
It is a pity that the adapter does not seem to provide _any_ (!) indication whatsoever on what is going on on the RS485 side... :(

I checked on the inverter (FoxESS H3 10.0) but also there I cannot find any indication on what is going on on the RS485 side. 
The cliend ID is configured to 247 as written in the howto.

Any idea on what I could do to further debug this chain?

**Comment by FozzieUK** [comment]
The 192.168.0.201 IP address was just given as an example of the configuration, if your network is in the 172.16.66.x range - you need to select a free IP address in that range and set it to that.

**Reply by ccbeorn** [reply]
That's the thing. I thought the same. But the IP field is a read only field, I cannot change its value. It only becomes changeable when I select TCP Client (to then hold the server IP).

I assigned a home-network-compliant IP in the network configuration and the device uses that correctly.
I'm just very confused about this TCP Server IP that doesn't fit but also cannot be changed.

**Comment by FozzieUK** [comment]
As you’ve got an H3 when you use the modpoll tool, you’ll need to change the -t 3 to -t 4 as the H3 has holding registers (not inputs) can you try polling with -t 4

**Reply by ccbeorn** [reply]
I then get a different error message:

-- polling slave...
Illegal Data Address exception response!

**Reply by canton7** [reply, marked as answer]
That implies you're trying to poll an incorrect register. It does mean you're talking to the inverter, so success!

For the H3, try 31000. 11000 is an input register, and isn't exposed by the H3.

**Reply by ccbeorn** [reply]
Wow, you're right! That way, I got the expected return!

```
G:\system_tools\modpoll\win>modpoll.exe -m enc -p 502 -a 247 -t 4 -0 -1 -c 16 -r 31000 172.16.66.118
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: Encapsulated RTU over TCP, FC3
Slave configuration...: address = 247, start reference = 31000 (PDU), count = 16
Communication.........: 172.16.66.118, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
[31000]: 4581
[31001]: 17
[31002]: 797
[31003]: 4546
[31004]: 0
[31005]: 40
[31006]: 2299
[31007]: 2291
[31008]: 2296
[31009]: 10
[31010]: 11
[31011]: 11
[31012]: 144
[31013]: 185
[31014]: 198
[31015]: 4998
```

And the values are in line with what the energy-stats app from AI priest is showing for my FoxEss system (considering the scale factor of the according register).

Perfect!! Thanks so much to you two! The -t 4 and the right register address did the trick!

**Reply by ccbeorn** [reply]
One thing I would need to understand in order to further make use of the modbus data:

the documentation of the modpoll tool states that -t 4 means that I am polling 16-bit from a holding register. With -r, I define the starting register.

In a general modbus documentation, I found the following statement:

> In the traditional standard, [entity] numbers for those entities start with a digit, followed by a number of four digits in range 1–9,999:
> 
> coils numbers start with a zero and then span from 00001 to 09999
> discrete input numbers start with a one and then span from 10001 to 19999
> input register numbers start with a three and then span from 30001 to 39999
> holding register numbers start with a four and then span from 40001 to 49999
> 
> This translates into [entity] addresses between 0 and 9,998 in data frames.

In the modbus register map for the FoxESS Inverter, I clearly need the 31xxx registers which are carrying the data I need. From the definition above, I would expect these to be input registers. But why do I need to poll with -t 4 for holding registers?

**Reply by canton7** [reply]
The H3 does not expose input registers.

See https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/Modbus-Map.xls for the H3 register map (select the appropriate tab)

**Comment by calum-mcfarlane** [comment]
Also check your protocol setting, you've got it set to **Encapsulated RTU over IP**, but I suspect you want **Modbus/TCP**.

**Reply by ccbeorn** [reply]
With - t 4 and - m tcp
I get a timeout again (contrary to the illegal data address exception if I use - t 4 with - m enc)

**Reply by canton7** [reply]
No, this is a TCP to RS485 adapter (no ModbusTCP), so `-m enc` is correct

**Reply by calum-mcfarlane** [reply]
Ah, I stand corrected, apologies for the bad advice OP!

**Comment by fapgomes** [comment]
Hi,

 I've the same problem with USR-TCP232-304. I've configured the USR like this:

https://github.com/nathanmarlor/foxess_modbus/wiki/USR-TCP232-304

But I've this error:

 
![Screenshot_20250519_192450](https://github.com/user-attachments/assets/357a804f-febe-44ea-97b6-3cbda4abdd2f)

and with modpoll I've this:

$  ./modpoll -m enc -p 502 -a 247 -t 3 -0 -1 -c 16 -r 11000 10.10.6.6
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: Encapsulated RTU over TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: 10.10.6.6, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
Checksum error!
$  ./modpoll -m enc -p 502 -a 247 -t 3 -0 -1 -c 16 -r 11000 10.10.6.6
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: Encapsulated RTU over TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: 10.10.6.6, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
Reply time-out!

I already tried with :

$  ./modpoll -m enc -p 502 -a 247 -t 4 -0 -1 -c 16 -r 11000 10.10.6.6
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: Encapsulated RTU over TCP, FC3
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: 10.10.6.6, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
Checksum error!
$  ./modpoll -m enc -p 502 -a 247 -t 4 -0 -1 -c 16 -r 31000 10.10.6.6
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: Encapsulated RTU over TCP, FC3
Slave configuration...: address = 247, start reference = 31000 (PDU), count = 16
Communication.........: 10.10.6.6, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
Reply time-out!

My USR-304 is like this:

![Screenshot_20250519_192641](https://github.com/user-attachments/assets/1eb739d9-e97f-43ac-8e83-f5bb3f08a038)

Thanks in advance,
